File: backend/app/application/ports/session_registration_repository.py
from abc import ABC, abstractmethod
from typing import Optional, List
from app.domain.entities.session_registration import SessionRegistration

class SessionRegistrationRepository(ABC):
    @abstractmethod
    def save(self, registration: SessionRegistration) -> SessionRegistration:
        pass

    @abstractmethod
    def get_by_user_and_session(self, user_id: int, session_id: int) -> Optional[SessionRegistration]:
        pass

    @abstractmethod
    def list_by_session(self, session_id: int) -> List[SessionRegistration]:
        pass

    @abstractmethod
    def delete(self, registration_id: int) -> None:
        pass

File: frontend/src/app/core/domain/ports/session-registration.repository.ts
import { Observable } from 'rxjs';
import { SessionRegistration } from '@core/domain/entities/session-registration.entity';

export abstract class SessionRegistrationRepository {
  abstract registerToSession(sessionId: number): Observable<SessionRegistration>;
  abstract unregisterFromSession(sessionId: number): Observable<void>;
  abstract getByUserId(userId: number): Observable<SessionRegistration[]>;
}
